crypto/bio/bio_lib.c

```diff
@@ -856,21 +863,21 @@ int BIO_connect_retry(BIO *bio, int timeout)
 
     if (rv <= 0) {
         if (get_last_sys_error() == ETIMEDOUT) {
-            /*
-             * if blocking, despite blocking BIO, BIO_do_handshake() timed out
-             * when non-blocking, BIO_do_handshake() timed out early
-             * with rv == -1 and get_last_sys_error() == 0
+            /*-
+             * If blocking, BIO_do_handshake() timed out despite blocking BIO.
+             * If non-blocking, BIO_do_handshake() timed out early with
+             *                  rv == -1 and get_last_sys_error() == 0.
              */
             ERR_clear_error();
             (void)BIO_reset(bio);
             /*
-             * unless using BIO_reset(), blocking next connect() may crash and
-             * non-blocking next BIO_do_handshake() will fail
+             * Unless using BIO_reset(), blocking next connect() may crash and
+             * non-blocking next BIO_do_handshake() would fail.
```

Comment:
I am a bit unsure if that is a symptom of a deeper problem.
Do you have some details what "crash" may have happened here?

Reply:
Unfortunately I don't recall the details.
It was about two years back when I had trouble implementing a stable HTTP connect
and asked the OpenSSL gurus for help (as far as I remember on the mailing list).
As I did not receive it I experimented until I found a working solution, as given here.

Reply:
I wouldn't block this PR for it, as it is just a range check, but this
explanation given in this comment raises more questions than it answers.
So probably need more investigations.

Reply:
Meanwhile I found the email exchange I referred to 3 days ago: 
https://mta.openssl.org/pipermail/openssl-dev/2017-September/009697.html

Reply:
The comments that - justifiably - raised questions are gone meanwhile,
due to the re-structuring of the retry checks that I did yesterday.
Are you fine with the new version?